man/tss2_authorizepolicy.1.md
% tss2_authorizepolicy(1) tpm2-tools | General Commands Manual
%
% APRIL 2019

# NAME

**tss2_authorizepolicy**(1) -

# SYNOPSIS

**tss2_authorizepolicy** [*OPTIONS*]

[common fapi references](common/tss2-fapi-references.md)

# DESCRIPTION

**tss2_authorizepolicy**(1) - This command signs a given policy with a given key such that the policy can be referenced from other policies that contain a corresponding PolicyAuthorize elements. The signature is done using the TPM signing schemes as specified in the cryptographic profile (cf., **fapi-profile(5)**).

# OPTIONS

These are the available options:

  * **-P**, **\--policyPath**=_STRING_:
    Path of the new policy.

    A policyPath is composed of two elements, separated by "/". A policyPath
    starts with "/policy". The second path element identifies the policy
    or policy template using a meaningful name.

  * **-p**, **\--keyPath**=_STRING_:
    Path of the signing key.

  * **-r**, **\--policyRef**=_FILENAME_ or _-_ (for stdin):
    A byte buffer to be included in the signature. Optional parameter.

[common tss2 options](common/tss2-options.md)

# EXAMPLE
```
tss2_authorizepolicy --keyPath=HS/SRK/myPolicySignKey --policyPath=/policy/pcr-policy --policyRef=policyRef.file
```

# RETURNS

0 on success or 1 on failure.

[footer](common/footer.md)


CHANGELOG
v2.2.2 - next:
*  fix make distcheck
v2.1.1 - next:
* Potential memory leak fix when tcti/sapi initialization fails.
* tpm2_listpcrs: use TPM2_GetCapability to determine PCRs to read
* listpcrs: remove one redundant call to tpm get cap
* listpcrs: fix for unsupported/disabled alg in -L
* build: use supported comment to suppress GCC7 fallthrough warning
* kdfa: allow to build with OpenSSL 1.1.x

v2.1.0 - 2017-7-24
* Fix readx and writex on multiple EINTR returns.
* Add support for the tabrmd TCTI. This is the new default.
* Change default socket port from 2323 (the old resourcemgr) to 2321 (default
  simulator port).
* Cherry-pick fix for CVE-2017-7524.
* Fix tpm2_listpcr command line option handling.
* Fix tpm2_getmanufec memory issues.

v2.0 - 2017-03-29
* Tracked on the milestone: https://github.com/01org/tpm2.0-tools/milestone/2
* Reworked all the tools to support configurable TCTIs, based on build time
  configuration, one can specify the tcti via the --tcti (-T) option to all
  tools.
* tpm2_getrandom interface made -s a positional argument.
* Numerous bug fixes.

v1.1 - 2016-11-04
Added
* travis ci support.
* Allow for unit tests to be enabled selectively.
* tpm2_rc_decode tool: Decode TPM_RC error codes.
* Android Make file
* tpm2_listpersistent: list all persistent objects
* test scripts for tpm2-tools
* tpm2_nvreadlock
* tpm2_getmanufec: retrieve EC from tpm manufacturer server.
* Copy 'common' and 'sample' code from the TPM2.0-TSS repo.

Modified
* tpm2_takeownership: update option -c to use lockout password to clear.
* tpm2_listpcrs: add options -L and -s, rewrite to increase performance.
* tpm2_quote: added -L option to support selection of multiple banks.
* tpm2_quote: add -q option to get qualifying data.
* configure: Use pkg-config to get info about libcurl and libcrypto.
* configure: Use pkg-config to locate SAPI and TCTI headers / libraries.
* tpm2_x: Add -X option to enable password input in Hex format.
* tpm2_nvdefine: Change -X option to -I.
* tpm2-nvwrite: fix for unable to write 1024B+ data.
* tpm2_getmanufec: Fix base64 encoding.
* tpm2_x: fixed a lot of TPM2B failures caused by wrong initialization.
* tpm2_getmanufec: let configure handle libs.
* tpm2_getmanufec: Convert from dos to unix format.
* build: Check for TSS2 library @ configure time.
* build: Detect required TSS2 and TCTI headers.
* build: Use libtool to build the common library
* build: Install all binaries into sbin.
* build: Build common sources into library.
* build: Move all source files to 'src'.
* Makefile.am: Move all build rules into single Makefile.am.
* everything: Use new TCTI headers and fixup API calls.
* everything: Update source to cope with sapi header cleanup.
* tpm2_activatecredential: Updated to support TCG compatible EK
* tpm2_getpubak: Updated to use TCG compatible EK
* tpm2_getpubek: fix ek creation to follow TCG EK profile spec.

Removed
* Windows related code
* depenedency on the TPM2.0-TSS repo source code

v1.0 - 2015-10-19
* 1.0 release
* 29 tools included


tools/fapi/tss2_authorizepolicy.c
/* SPDX-License-Identifier: BSD-3-Clause */

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "tools/fapi/tss2_template.h"

/* Context struct used to store passed commandline parameters */
static struct cxt {
    char  const *policyPath;
    char  const *keyPath; /* the path to the signing key */
    char  const *policyRef;
} ctx;

/* Parse commandline parameters */
static bool on_option(char key, char *value) {
    switch (key) {
    case 'P':
        ctx.policyPath = value;
        break;
    case 'p':
        ctx.keyPath = value;
        break;
    case 'r':
        ctx.policyRef = value;
        break;
    }
    return true;
}

/* Define possible commandline parameters */
static bool tss2_tool_onstart(tpm2_options **opts) {
    struct option topts[] = {
        {"policyPath", required_argument, NULL, 'P'},
        {"keyPath",    required_argument, NULL, 'p'},
        {"policyRef",    required_argument, NULL, 'r'},
    };
    return (*opts = tpm2_options_new ("P:p:r.", ARRAY_LEN(topts), topts,
                                      on_option, NULL, 0)) != NULL;
}

/* Execute specific tool */
static int tss2_tool_onrun (FAPI_CONTEXT *fctx) {
    /* Check availability of required parameters */
    if (!ctx.policyPath) {
        fprintf (stderr, "policy path to sign is missing, pass" \
            "--policyPath\n");
        return -1;
    }
    if (!ctx.keyPath) {
        fprintf (stderr, "key path for signing key is missing, pass" \
            "--keyPath\n");
        return -1;
    }

    /* Read ciphertext file */
    TSS2_RC r;
    uint8_t *policyRef = NULL;
    size_t policyRefSize = 0;
    if (ctx.policyRef){
        r = open_read_and_close (ctx.policyRef, (void**)&policyRef,
            &policyRefSize);
        if (r){
            return r;
        }
    }

    /* Execute FAPI command with passed arguments */
    r = Fapi_AuthorizePolicy (fctx, ctx.policyPath, ctx.keyPath, policyRef,
        policyRefSize);
    if (r != TSS2_RC_SUCCESS){
        LOG_PERR ("Fapi_AuthorizePolicy", r);
        free (policyRef);
        return r;
    }

    free (policyRef);

    return r;
}

TSS2_TOOL_REGISTER("authorizepolicy", tss2_tool_onstart, tss2_tool_onrun, NULL)
